packages/core/src/di/r3_injector.ts

```diff
@@ -95,35 +101,203 @@ export function createInjector(
+        const {ngModule, providers} = importTypesWithProviders[i];
+        deepForEach(providers!, provider => {
+          validateProvider(provider, providers || EMPTY_ARRAY, ngModule);
+          providersOut.push(provider);
+        });
+      }
+    }
+  }
+  // Track the InjectorType and add a provider for it.
+  // It's important that this is done after the def's imports.
+  const factory = getFactoryDef(defType) || (() => new defType());
+
+  // Provider to create `defType` using its factory.
+  providersOut.push({
+    provide: defType,
+    useFactory: factory,
+    deps: EMPTY_ARRAY,
+  });
+
+  providersOut.push({
+    provide: INJECTOR_DEF_TYPES,
+    useValue: defType,
+    multi: true,
+  });
+
+  // Provider to eagerly instantiate `defType` via `INJECTOR_INITIALIZER`.
+  providersOut.push({
+    provide: INJECTOR_INITIALIZER,
+    useValue: () => inject(defType),
+    multi: true,
+  });
+
+  // Next, include providers listed on the definition itself.
+  const defProviders = def.providers;
+  if (defProviders != null && !isDuplicate) {
+    const injectorType = container as InjectorType<any>;
+    deepForEach(defProviders, provider => {
+      // TODO: fix cast
+      validateProvider(provider, defProviders as any[], injectorType);
+      providersOut.push(provider);
+    });
+  }
+
+  return (
+      ngModule !== undefined &&
+      (container as InjectorTypeWithProviders<any>).providers !== undefined);
+}
+
+
+
+/**
+ * An `Injector` that's part of the environment injector hierarchy, which exists outside of the
+ * component tree.
+ */
+export abstract class EnvironmentInjector implements Injector {
   /**
-   * Map of tokens to records which contain the instances of those tokens.
-   * - `null` value implies that we don't have the record. Used by tree-shakable injectors
-   * to prevent further searches.
+   * Retrieves an instance from the injector based on the provided token.
+   * @returns The instance from the injector if defined, otherwise the `notFoundValue`.
+   * @throws When the `notFoundValue` is `undefined` or `Injector.THROW_IF_NOT_FOUND`.
    */
-  private records = new Map<ProviderToken<any>, Record<any>|null>();
-
+  abstract get<T>(token: ProviderToken<T>, notFoundValue?: T, flags?: InjectFlags): T;
   /**
-   * The transitive set of `InjectorType`s which define this injector.
+   * @deprecated from v4.0.0 use ProviderToken<T>
+   * @suppress {duplicate}
    */
-  private injectorDefTypes = new Set<InjectorType<any>>();
+  abstract get(token: any, notFoundValue?: any): any;
+
+  abstract destroy(): void;
+
+  abstract onDestroy(callback: () => void): void;
+}
+
+/**
+ * Collects providers from all NgModules, including transitively imported ones.
+ *
+ * @returns The list of collected providers from the specified list of NgModules.
+ * @publicApi
+ */
+export function importProvidersFrom(...injectorTypes: Array<Type<unknown>>): Provider[] {
+  const providers: SingleProvider[] = [];
```

Comment:
```suggestion
  const providers: Provider[] = [];
```
(so that it's aligned with the return type)

Reply:
`SingleProvider` doesn't allow recursion (nested provider arrays). It's a good distinction to have here.